GraphQL Basics [hw from 18.11.2021]

RasimKhusaenov opened the pull request:
It's not completely ready yet. There is a problem with the CreateTask mutation — it crashes with an error.

RasimKhusaenov committed:
Add params for FilteredTasksQuery

RasimKhusaenov committed:
Add interactor to update project

RasimKhusaenov committed:
Add sending notification to update project

RasimKhusaenov committed:
Fix update condition

RasimKhusaenov committed:
Add UpdateProject mutation

RasimKhusaenov committed:
Add DeleteProject mutation

RasimKhusaenov committed:
Add CommentType

RasimKhusaenov committed:
Add interactors for Task methods

RasimKhusaenov committed:
Add CreateTask mutation
It fails with error:

app/graphql/mutations/create_task.rb:22:in `resolve'
wrong number of arguments (given 2, expected 1)

RasimKhusaenov committed:
Add API for all Task methods

RasimKhusaenov committed:
Add interactors for all Comment methods

RasimKhusaenov committed:
Add API for all Comment methods

RasimKhusaenov committed:
Add API for all Comment methods

RasimKhusaenov committed:
Merge remote-tracking branch 'origin/graphql-basics' into graphql-basics

DmitryBarskov commented on app/controllers/comments_controller.rb:
```diff
@@ -3,28 +3,38 @@ class CommentsController < ApplicationController
 
   def create
-    comment = Comment.new(comment_params)
+    @comment = create_comment.comment
```
Why do you use an instance variable here? Do you understand that they won't be available in TasksController?

RasimKhusaenov replied:
No, unfortunately, I don't understand. (//ω//)

RasimKhusaenov committed:
Fix queries for Task
It was impossible to view tasks and task via the API

RasimKhusaenov committed:
Fix CommentType
Instead of one task, the field asked for several

RasimKhusaenov committed:
Fix comments
The code from the last hw has returned and the display of comments in the task has been fixed

DmitryBarskov commented on app/controllers/comments_controller.rb:
```diff
@@ -34,7 +45,8 @@ def comment_params
         {
-          user: current_user
+          user: current_user,
+          task_id: params[:task_id]
```
why? https://github.com/DmitryBarskov/task-tracker/pull/64/files#diff-34e52388abd1ae0658edd56e12cd6d935fac98673361084ba07a99ca4bb836e7R45

AlexanderTheThird replied:
ID is already in params, isn't it?

RasimKhusaenov replied:
Without this, the comment has no task_id

DmitryBarskov replied:
eliminate duplication of params

DmitryBarskov commented on app/controllers/tasks_controller.rb:
```diff
@@ -1,22 +1,26 @@
 class TasksController < ApplicationController
   before_action :set_task, only: %i[show edit update destroy]
+  before_action :set_task, only: %i[show edit update destroy]
```
why? https://github.com/DmitryBarskov/task-tracker/pull/64/files#diff-de1aaaf7a00c963e6a5e77045d8cbdafa5ebc81e1250a7ea1862a51d824e9945R2

DmitryBarskov commented on app/policies/project_policy.rb:
```diff
@@ -14,7 +14,7 @@ def create?
 
   def update?
-    user.id == project.user_id
+    user.id.in? project.user_ids
```
Why do you use this method?

RasimKhusaenov replied:
This is from an old homework. Because we have several users who can participate in one project, so everyone should be able to edit the project.

DmitryBarskov commented on app/query_objects/filtered_tasks_query.rb:
```diff
@@ -8,4 +8,12 @@ def by_status(relation, status)
+  end
+
+  def by_deadline_at(relation, deadline)
+    relation.where(deadline_at: deadline)
```
It has no sense, since it must be equal even by seconds

RasimKhusaenov replied:
It will be better if I implement it as follows:
```suggestion
    relation.where(deadline_at: deadline.midnight..(deadline.midnight + 1.day))
```
?

DmitryBarskov replied:
Do you even use this method somewhere?

RasimKhusaenov committed:
Remove duplicate line

RasimKhusaenov committed:
Fix update method and authorization
